### PR TITLE
Unit test user management

### PR DESCRIPTION
**Pull request by kohkaixun**
Under Private.test.js, the "it should render Spinner if axios request fails" (5th test) keeps throwing an error. Private.js doesn't have proper error validation to exit gracefully and that might be the issue. I feel like the test should be able to run and then show that test fails, however, whenever I run the test, the test doesn't seem to even run and I think the new error I created is thrown.

### CODE REVIEW

File: client/src/components/Routes/AdminRoute.test.js [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+import React from 'react';
+import { render, screen } from '@testing-library/react';
+import Spinner from "../Spinner";
+import { useAuth } from '../../context/auth';
+
+jest.mock('../../context/auth', () => ({
+    useAuth: jest.fn()
+}));
+
+jest.mock('mongoose', () => ({
```
Comment (eugenetangkj) [inline client/src/components/Routes/AdminRoute.test.js]: Why is `mongoose` being mocked? I dont see it being used in `AdminRoute.js`

File: client/src/components/Routes/AdminRoute.test.js [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+);
+
+describe('AdminRoute', () => {
+    beforeEach(() => {
+        jest.clearAllMocks();
+    });
+
+    it('should render Spinner if no auth token', async () => {
+        useAuth.mockReturnValue([null, jest.fn()]);
+        render(<AdminRoute />);
```
Comment (eugenetangkj) [inline client/src/components/Routes/AdminRoute.test.js]: Do you want to expect that `axios.get` is called with a particular argument, for example `/api/v1/auth/admin-auth`?

File: client/src/components/Routes/Private.test.js [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+import React from 'react';
+import { render, screen } from '@testing-library/react';
+import Spinner from "../Spinner";
+import { useAuth } from '../../context/auth';
+
+jest.mock('../../context/auth', () => ({
+    useAuth: jest.fn()
+}));
+
+jest.mock('mongoose', () => ({
```
Comment (eugenetangkj) [inline client/src/components/Routes/Private.test.js]: Again, why is `mongoose` mocked?

File: controllers/authController.updateProfileController.test.js [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+
+        let updatedUser = JSON.parse(JSON.stringify(newProfileUpdate));
+        updatedUser.password = passwordHash;
+
+        await updateProfileController(req, res);
+
+        expect(userModel.findByIdAndUpdate).toHaveBeenCalledTimes(1);
+        expect(userModel.findByIdAndUpdate).toHaveBeenCalledWith(req.user._id, updatedUser, { new: true });
+        expect(res.status).toHaveBeenCalledWith(200);
+        expect(res.send).toHaveBeenCalledTimes(1);
```
Comment (eugenetangkj) [inline controllers/authController.updateProfileController.test.js]: Not sure if you want to assert the content of the response, such as `success` and `message` attributes.

File: controllers/authController.updateProfileController.test.js [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+            address: "",
+            phone: ""
+        };
+
+        let updatedUser = JSON.parse(JSON.stringify(oldProfileUpdate));
+
+        await updateProfileController(req, res);
+        expect(userModel.findByIdAndUpdate).toHaveBeenCalledTimes(1);
+        expect(userModel.findByIdAndUpdate).toHaveBeenCalledWith(req.user._id, updatedUser, { new: true });
+        expect(res.status).toHaveBeenCalledWith(200);
```
Comment (eugenetangkj) [inline controllers/authController.updateProfileController.test.js]: Your test is testing 'Should keep old profile if new profile is empty'. However, your `expect` statements do not seem to be verifying whether the contents of the user are being changed or not.

File: controllers/authController.updateProfileController.test.js [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+        expect(res.send).toHaveBeenCalledTimes(1);
+    });
+
+    // This test will fail as there is currently no validation for phone number in original code.
+    it("should not update profile with invalid phone number", async () => {
```
Comment (eugenetangkj) [inline controllers/authController.updateProfileController.test.js]: Would be good to leave a comment above the tests that you do not expect to pass so it's easier for us to read! 😄

**Review by eugenetangkj — COMMENTED** [review]
Great work so far! Keep it up

File: controllers/authController.updateProfileController.test.js (hunk 4 above)
Comment (kohkaixun) [reply controllers/authController.updateProfileController.test.js]: Will that make the test too brittle?

File: controllers/authController.updateProfileController.test.js (hunk 5 above)
Comment (kohkaixun) [reply controllers/authController.updateProfileController.test.js]: I think updatedUser declared in line 182 is made from parsing the oldProfileUpdate declared in line 15. I just change the name of the updatedUser to oldProfileUpdate? Or is there some other logic error that I'm missing?

File: client/src/pages/user/Dashboard.test.js [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        expect(screen.getByText("User Menu")).toBeInTheDocument();
+        expect(screen.getByText(name)).toBeInTheDocument();
+        expect(screen.getByText(email)).toBeInTheDocument();
+        expect(screen.getByText(address)).toBeInTheDocument();
+    });
+
+    it("should not render user information if auth is null", () => {
+        useAuth.mockReturnValue([null]);
+        render(<Dashboard />);
+        expect(screen.queryByText(name)).not.toBeInTheDocument();
```
Comment (eugenetangkj) [inline client/src/pages/user/Dashboard.test.js:35]: Hmmm... okay, but maybe another approach would have been to check that the h3 elements are not rendered with any content.

File: client/src/pages/user/Profile.test.js [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+        const passwordInput = screen.getByPlaceholderText(passwordInputPlaceholderText);
+        const phoneInput = screen.getByPlaceholderText(phoneInputPlaceholderText);
+        const addressInput = screen.getByPlaceholderText(addressInputPlaceholderText);
+
+        fireEvent.change(nameInput, { target: { value: updatedUserProfile.user.name } });
+        fireEvent.change(passwordInput, { target: { value: updatedUserProfile.user.password } });
+        fireEvent.change(phoneInput, { target: { value: updatedUserProfile.user.phone } });
+        fireEvent.change(addressInput, { target: { value: updatedUserProfile.user.address } });
+
+        expect(nameInput.value).toBe(updatedUserProfile.user.name);
```
Comment (eugenetangkj) [inline client/src/pages/user/Profile.test.js:103]: Hmmm, this is checking whether the fields is editable + whether the updating of profile is correct. -> 2 different things.

But okay for this round ba

**Review by eugenetangkj — APPROVED** [review]
Thanks! Looks good, left some comments for your reference.

**Pull request merged**